src/buffer/line_cache.rs

```diff
@@ -0,0 +1,115 @@
+use std::ops::Range;
+
+use crate::{document::ReadableDocument, simd::memchr2};
+
+/// Cache a line/offset pair every CACHE_EVERY lines to speed up line/offset calculations
+const CACHE_EVERY: usize = 1024 * 64;
+
+pub struct LineCache {
+    cache: Vec<(usize, usize)>, // (index, line)
```

Comment:
Instead of storing indices/lines, ideally, we would store index/line _deltas_ between items. During updates, we then only need to update the items that are affected by the change, but not all items past it.